src/middleware.js

```diff
@@ -0,0 +1,25 @@
+import { NextResponse } from "next/server";
+import { session } from "./lib/server-session";
+import * as jose from "jose";
+
+export async function middleware(request) {
+  const { token } = session();
+  const secretKey = process.env.SECRET_KEY;
+  const secret = new TextEncoder().encode(secretKey);
+
+  if (!token) {
+    return NextResponse.redirect(new URL("/login", request.url));
+  }
+
+  //validating token
+  try {
+    await jose.jwtVerify(token, secret);
+    return NextResponse.next();
+  } catch (error) {
+    return NextResponse.redirect(new URL("/login", request.url));
+  }
+}
+
+export const config = {
+  matcher: ["/dashboard"],
+};
```

Comment:
We might need to handle cases for other urls too. This is not the preferred solution.
The matcher should match all pages route (not static assets, i.e the one ends with .js, .css).